subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/CurrentTaskExecution.java

```diff
@@ -0,0 +1,106 @@
+    private final ImmutableSortedMap<String, FileCollectionSnapshot> inputFilesSnapshot;
+    private FileCollectionSnapshot discoveredInputFilesSnapshot;
+    private final OverlappingOutputs detectedOverlappingOutputs;
+    private Boolean successful;
+
+    public CurrentTaskExecution(
+        UniqueId buildInvocationId,
+        ImplementationSnapshot taskImplementation,
+        ImmutableList<ImplementationSnapshot> taskActionImplementations,
+        ImmutableSortedMap<String, ValueSnapshot> inputProperties,
+        ImmutableSortedSet<String> outputPropertyNames,
+        ImmutableSet<String> declaredOutputFilePaths,
+        @Nullable OverlappingOutputs detectedOverlappingOutputs,
+        ImmutableSortedMap<String, FileCollectionSnapshot> inputFilesSnapshot,
+        FileCollectionSnapshot discoveredInputFilesSnapshot,
+        ImmutableSortedMap<String, FileCollectionSnapshot> outputFilesSnapshot
+    ) {
+        super(buildInvocationId, taskImplementation, taskActionImplementations, inputProperties, outputPropertyNames, declaredOutputFilePaths);
+        this.detectedOverlappingOutputs = detectedOverlappingOutputs;
+        this.outputFilesSnapshot = outputFilesSnapshot;
+        this.inputFilesSnapshot = inputFilesSnapshot;
+        this.discoveredInputFilesSnapshot = discoveredInputFilesSnapshot;
+    }
+
+    public boolean isSuccessful() {
+        return successful;
+    }
+
+    public void setSuccessful(boolean successful) {
+        this.successful = successful;
+    }
+
+    @Override
+    public ImmutableSortedMap<String, FileCollectionSnapshot> getOutputFilesSnapshot() {
+        return outputFilesSnapshot;
+    }
+
+    public void setOutputFilesSnapshot(ImmutableSortedMap<String, FileCollectionSnapshot> outputFilesSnapshot) {
+        this.outputFilesSnapshot = outputFilesSnapshot;
+    }
+
+    @Override
+    public ImmutableSortedMap<String, FileCollectionSnapshot> getInputFilesSnapshot() {
+        return inputFilesSnapshot;
+    }
+
+    @Override
+    public FileCollectionSnapshot getDiscoveredInputFilesSnapshot() {
+        return discoveredInputFilesSnapshot;
+    }
+
+    public void setDiscoveredInputFilesSnapshot(FileCollectionSnapshot discoveredInputFilesSnapshot) {
+        this.discoveredInputFilesSnapshot = discoveredInputFilesSnapshot;
+    }
+
+    @Nullable
+    public OverlappingOutputs getDetectedOverlappingOutputs() {
+        return detectedOverlappingOutputs;
+    }
+
+    public TaskExecutionSnapshot snapshot() {
+        return new TaskExecutionSnapshot(
+            successful,
+            getBuildInvocationId(),
+            getTaskImplementation(),
+            getTaskActionImplementations(),
+            getOutputPropertyNamesForCacheKey(),
+            getDeclaredOutputFilePaths(),
+            getInputProperties(),
+            inputFilesSnapshot,
+            discoveredInputFilesSnapshot,
+            outputFilesSnapshot
+        );
```

Comment:
Is there maybe a way to get rid of the notion of a `TaskExecutionSnapshot`, and just have `TaskExecution`s? Creating the snapshot object seems like a waste to me.

Reply:
See my [comment](https://github.com/gradle/gradle/pull/3036#discussion_r141308459) above. We can get rid of `TaskExecutionSnapshot` completely. I think we are not creating much waste since we only create one new object which has references to already present objects - so we do not copy anything.